### DISCUSSION TITLE
Unable to resolve service "symfony_cli" to a factory during update

### DISCUSSION THREAD

**Discussion by Pyr0technicien** [monarc-project/MonarcAppFO#451] [General]
Hello,

i'm trying to update our version of monarc to the latest but i'm encoutering this error : 

![image](https://user-images.githubusercontent.com/14542651/181788891-3e45d71b-1598-47c5-9180-e8fa286b03af.png)

Any idea of what i'm doing wrong ?

**Comment by ruslanbaidan** [comment]
Hello,

Could you describe the steps you do to perform the update.

**Comment by Pyr0technicien** [comment]
I'm just changing the owner of all the folders : sudo chown -R myuser . 
As the user of the folder is www-data, then i launch the update script : 

./scripts/update-all.sh -c

**Comment by ruslanbaidan** [comment]
Could you try without the option "-c".

**Comment by Pyr0technicien** [comment]
I'm already using the option -c

**Comment by ruslanbaidan** [comment]
you can clear cache manually, could be an issue with permission as before the data was created by www-data.

**Comment by Pyr0technicien** [comment]
It goes a bit further but I'm still stuck at the end :

![image](https://user-images.githubusercontent.com/14542651/182105670-4bd6cd5d-5c7a-4997-8ae4-3ff316a6b59a.png)

When I remove the cache & restart apache2, the main page is loading but as soon as i want to login with my user, it loads for ever and never logs me in. 


These are the errors I get in my browser console : 

![image](https://user-images.githubusercontent.com/14542651/182106683-3cc173af-ad24-4638-bf94-89a89df9f8c6.png)

**Comment by ruslanbaidan** [comment]
The error message on the screen don't provide any useful information.
It only tells that there is another argument was passed to the script that is not supported, but it supposed to be run without any arguments.

The login problems could be related to missing migrations. You can try to run upgrade-db.sh

Regarding error messages: we need browser console Response tab errors.

**Reply by Pyr0technicien** [reply]
I edited my response with the browser console.

The argument error is coming from a "laminas" command in the update script, not from the arguments passed to the script

**Reply by ruslanbaidan** [reply]
Browser Network tab Response. Previously was written not correctly. With the red line in focus.
![Screenshot 2022-08-01 at 10 48 24](https://user-images.githubusercontent.com/3246171/182110953-177d6e29-b096-4bb8-9476-68793ebfc86c.png)

**Reply by ruslanbaidan** [reply]
There are no laminas commands run when you run the update-all.sh script without arguments.

**Reply by Pyr0technicien** [reply]
I run it with the arguments " -c " as mentionned ealier

**Reply by Pyr0technicien** [reply]
![image](https://user-images.githubusercontent.com/14542651/182114560-b4c5b069-6b9c-4679-ba8e-a7014f366f14.png)

That's the last thing I can see

**Reply by ruslanbaidan** [reply]
It means you did not have any login issues.
If the Monarc UI is not loaded then the issue is another. Search for red line in the list.
Have you run "upgrade-db.sh" after pulling the repo (git pull) and reinstalling all the dependencies (composer ins) ?

**Comment by Pyr0technicien** [comment]
So I restored a backup and did the process all over again.
I'm having the same issue : 
![image](https://user-images.githubusercontent.com/14542651/182110906-1107ebdd-471a-42c4-81ec-ea18978781dc.png)


But then I'm allowed to login.
However the monarc version is still 2.11.1 so the update didn't complete

**Reply by ruslanbaidan** [reply]
Did you do "git pull" of the repository ?

**Reply by Pyr0technicien** [reply]
Git pull is up to date, the update script does it for me

**Reply by ruslanbaidan** [reply]
If you see the version 2.11.1 then it could fail.
you can "cat package.json" version is stored there.
If it is not 12.2.2 then there is an issue.

**Comment by Pyr0technicien** [comment]
@ruslanbaidan, Ok well your messages helped me a lot.
I restoreed my VM to a previous state and ran as follow :

Delete cache
Re-Create cache folder
change owner of all files 
run : ./scripts/update-all.sh (no args)
Change back owner to www-data
restart apache2

So definitely the arg " -c " that was failing everything for some reason.

Now the version is correct v 2.12.2.

**Comment by ruslanbaidan** [comment]
Good news.
Hope Monarc works for you now ?

**Reply by Pyr0technicien** [reply]
Yes thanks, this upgrade was more difficult than expected but it's all good now

**Reply by ruslanbaidan** [reply]
I think the reason was with the permissions. Perhaps some directories or files did not have sufficient access to make "git pull", install dependencies and clean cache.

**Reply by Pyr0technicien** [reply]
Might have been the problem indeed, if that can maybe help someone : 

The official guide says 
"rm -Rf ./data/cache"

But this will erase the folder "cache" itself too, that will make the update script fail because the folder doesn't exist.

When you create that folder back, you need to be sure not to use sudo and you need to be sure to change the owner back to www-data after running the script.
I might have missed some steps everytime i ran it again. I ended up running the update script without "-c" argument so lamina stops failing.

I'm suprise the guide doesn't say anything about changing the owner of the folders before running the script, it would fail for me if I didn't change that. Also composer tells me I shouldn't run it as "sudo".

**Reply by ruslanbaidan** [reply]
The list of command to clean cache:
```
rm -rf ./data/cache/*
rm -rf ./data/DoctrineORMModule/Proxy/*
rm -rf ./data/LazyServices/Proxy/*
```
"composer ins -o --no-dev" should not be run with sudo, but the user should have write access to "vendor" and "bin" directories.

**Reply by Pyr0technicien** [reply]
Not sure where I was suppose to see all of this, I can't read anything about composer in the official documentation, therefore I never ran it.

I'm following the guide (literally has 3 command to update): 
https://www.monarc.lu/documentation/technical-guide/#monarc-update

The script is failling if all the folders are owned by www-data (git error if I remember correctly)

**Reply by ruslanbaidan** [reply]
The steps are described in [the installation guide](https://github.com/monarc-project/MonarcAppFO/blob/master/INSTALL/INSTALL.ubuntu1804.md#21-monarc-source-code),
including composer install. `--no-dev` is a preferred param to avoid installing dev dependencies.

The other approach with deleting and recreating of the cache directories defined in [the ansible playbook](https://github.com/monarc-project/ansible-ubuntu/blob/master/playbook/monarcfo/tasks/main.yaml#L163-L233).
Ansible is used to manage update for multiple clients.

**Comment by Pyr0technicien** [comment]
Well, hapiness was short, we can't connect to our instance anymore (only one account can).

For all other account we have the same issue as previously, when we click on connect, the loading ring never stops and we never get to enter monarc. no logs, nothing to be seen in the browser developer tools.

I performed the update-db.sh too but that doesn't solve the problem either. Any help would be appreciated as we are stuck here @ruslanbaidan

**Reply by ruslanbaidan** [reply]
Every clock on login button initiates an XmlHttpRequest from browser javascript script (Angular app) to the server. It should be listed on the Network XHR tab (and All by default). Perhaps you limited the time scale, try to refresh, open and close the dev tools.

You can also check the apache error logs (see you the vhost config to know where it is).

**Reply by Pyr0technicien** [reply]
When I say "nothing in developer tools", it means nothing unusual as we already tried to debug yesterday :
![image](https://user-images.githubusercontent.com/14542651/182564240-d0478299-0e04-45ba-9ccd-50b64692ecd3.png)

When I do have an errror with a cache folder not existing or something, I can see it in /var/log/apache2/error.log but here there is nothing, the page is just in an infinite loading loop

**Reply by ruslanbaidan** [reply]
You can try to clean the browser cache (Application tab on your screen), refresh with cache cleanup or open in another incognito tab or browser.

**Reply by Pyr0technicien** [reply]
All my test are done in a new incognito window, my cache isn't the problem.

Weird part is that one of our client can login.

**Reply by ruslanbaidan** [reply]
Are there any issues on the Console tab ?
Please clean it and click login again.

**Reply by Pyr0technicien** [reply]
I have this error before logging in
![image](https://user-images.githubusercontent.com/14542651/182568347-04f24e9a-78d9-4c18-b0b5-47525318ae52.png)

And then I have a second one when I click on log in 
![image](https://user-images.githubusercontent.com/14542651/182568479-6d6fb060-5c7c-488c-a1fe-1432800e27c9.png)

**Reply by ruslanbaidan** [reply]
That's exactly the same issue what was before fr the other user. The problem is the outdated front-end / Javascript files loaded.
Needs to be ensured:
1. Monarc FrontOffice instance is the same as for another client. If no, then it has to be updated.
2. The cache clean up was done properly

Based on the login response the client was successfully logged in, but the Javascript could not process the config parameter of 2 factor auth (the new feature).

**Reply by Pyr0technicien** [reply]
Yes, this only happens if the username / password is correct so indeed the login is sucessful.
When I said it was working for an other client, I meant for an other user account, same front office instance.

The cache is cleaned up properly, I'm opening a new incognito web browser each time, I tried on two different browser but it's the same issue.

The weird part is that my accounts were working properly two days ago when I updated monarc, but today this is happening again, without anybody touching the instance

EDIT : Our WAF might be the root cause for this, i'll have a look with the person responsible of it

EDIT 2  : Same problem without going through our waf unfortunatly

**Reply by ruslanbaidan** [reply]
Reinstalling of the frontend (JS) packages should help then.
Either they were reverted either not installed originally, but were used cached versions of JS.
Can be updated manually:
`cd node_modules/ng_anr`
`git pull`
`cd ../ng_client`
`git pull`
`npm install`

**Reply by Pyr0technicien** [reply]
When I run npm install, the ouput is pretty similar than when I updated to 1.12 : 

![image](https://user-images.githubusercontent.com/14542651/182579867-01afe642-e410-4329-8d14-b2cb0daf9e4e.png)

audit fix --force doesn't seem to solve anything : 

![image](https://user-images.githubusercontent.com/14542651/182581300-99eebb7f-478c-48e0-b242-ee92d422da66.png)
![image](https://user-images.githubusercontent.com/14542651/182581397-ddaf384c-13d2-47d7-867d-fafa419d187c.png)

Probelm still persist, but I guess because of this output, there is a problem with the frontend installation ?

**Reply by ruslanbaidan** [reply]
It seems there is an issue with external packages. Or perhaps the ng_client repo is quite outdated.
The easiest way to reinstall it entirely. When you are in {project-root}/node_modules folder:
`rm -rf ./ng_client`
`mkdir ng_client`
`git clone https://github.com/monarc-project/ng-client.git ./ng_client`
`cd ng_client`
`npm install`

**Reply by Pyr0technicien** [reply]
That's the result of the last "nmp install" of your list of commands

![image](https://user-images.githubusercontent.com/14542651/182584205-3f311d65-7462-44d1-948f-7b6c512f5c84.png)

And still same issue at login, How is it possible that for some user the log in process works though ?

**Reply by Pyr0technicien** [reply]
![image](https://user-images.githubusercontent.com/14542651/182591092-211d244e-0b4d-47ae-be91-16db50135d1f.png)
When I try to fix them

**Reply by ruslanbaidan** [reply]
> And still same issue at login, How is it possible that for some user the log in process works though ?
They probably have cached javascript.

Regarding the installation issue. Could you try to upgrade it:
`npm install -g npm@latest`

**Reply by Pyr0technicien** [reply]
Updated from 8.1 to 8.15 but still the same issue on npm install : 

![image](https://user-images.githubusercontent.com/14542651/182598196-912fa2a1-0e90-4a5c-9977-47e1a0531d06.png)


But these message don'tsay anything about the installation failing, just that there is some vulnerabilities in depedencies isn't it ?

**Reply by ruslanbaidan** [reply]
Angularjs library is used in Monarc has sub dependencies that are a bit outdated and we can't solve those issues on our side.
We don't use the vulnerable packages in the application.

If clients clean cache now Monarc should work.

**Reply by Pyr0technicien** [reply]
I'm not sure why it would be working now, nothing has changed, I still have the same error.

I still have no idea why I'm able to login with one of the users but every other users have the same error... 
Is this related to some user settings ?

I guess I will have to restore a backup for the third time and stop updating monarc, it seems I'm the only user having issue with this update ?

**Reply by Pyr0technicien** [reply]
I asked the only user that was still able to connect to clear its cache and now they have the same issue as all other users...

**Reply by ruslanbaidan** [reply]
Could there be a proxy cache configured ?

**Reply by ruslanbaidan** [reply]
What is the output of the `git status` command when you are in `node_modules/ng_client` and `node_modules/ng_anr` ?
If there is a tag or some branch, you need to execute `git checkout master` and `git pull`.

**Reply by Pyr0technicien** [reply]
No there is a waf in between but it's not caching anything. Also the issue is the same if we access directly monarc

**Reply by ruslanbaidan** [reply]
If the previous advise doesn't work, it would be better to reinstall the application.
To do this simply extract config from config/autoload/local.php somewhere, install the app and copy the config back.
Can be created a copy of the whole Monarc application folder before.
You only need to follow point 2.1
https://github.com/monarc-project/MonarcAppFO/blob/master/INSTALL/INSTALL.ubuntu1804.md#21-monarc-source-code

**Reply by Pyr0technicien** [reply]
Thanks for the time you spent with me on this.

I rolled back to 2.11 with our backup. We were assessing the need of keeping monarc on prem, we are going to decide if we go for the SAS option, if that's the case I won't need to update the application anymore. 

Thanks anyway

**Reply by ruslanbaidan** [reply]
Welcome.
The new installation normally takes 15 - 20 minutes. You don't need to configure the db, make any other system packages installation. The step 2.1 includes only the repos cloning, installing the dependencies (execute update-all.sh) creation of the data cache folders and the config copy from the current app.
So it's faster than downgrade and on some servers we perform upgrade by creating a new Monarc folder and just switching symlinks (the same what was suggested to you).
